### ISSUE TITLE
App Crashes in BraintreeSharedPreferences.java

### ISSUE THREAD

**Issue by tanmayyy** [braintree/braintree_android#619]
### **General information**
Environment: Production
Android Version and Device: Android 8.1.0, Huawei P20
Braintree dependencies:

```
implementation "com.braintreepayments.api:paypal:4.12.0
```

### Issue description
The app is crashing for this one particular user when he switches to the BrainTree component to complete the payment.

```
Fatal Exception: java.lang.SecurityException: Could not decrypt value. decryption failed
       at androidx.security.crypto.EncryptedSharedPreferences.getDecryptedObject(EncryptedSharedPreferences.java:579)
       at androidx.security.crypto.EncryptedSharedPreferences.getLong(EncryptedSharedPreferences.java:436)
       at com.braintreepayments.api.BraintreeSharedPreferences.getLong(BraintreeSharedPreferences.java:95)
       at com.braintreepayments.api.ConfigurationCache.getConfiguration(ConfigurationCache.java:41)
       at com.braintreepayments.api.ConfigurationCache.getConfiguration(ConfigurationCache.java:34)
       at com.braintreepayments.api.ConfigurationLoader.getCachedConfiguration(ConfigurationLoader.java:75)
       at com.braintreepayments.api.ConfigurationLoader.loadConfiguration(ConfigurationLoader.java:39)
       at com.braintreepayments.api.BraintreeClient$1.onAuthorizationResult(BraintreeClient.java:165)
       at com.braintreepayments.api.AuthorizationLoader.loadAuthorization(AuthorizationLoader.java:20)
       at com.braintreepayments.api.BraintreeClient.getAuthorization(BraintreeClient.java:174)
       at com.braintreepayments.api.BraintreeClient.getConfiguration(BraintreeClient.java:161)
       at com.braintreepayments.api.BraintreeClient$4.onAuthorizationResult(BraintreeClient.java:222)
       at com.braintreepayments.api.AuthorizationLoader.loadAuthorization(AuthorizationLoader.java:20)
       at com.braintreepayments.api.BraintreeClient.getAuthorization(BraintreeClient.java:174)
       at com.braintreepayments.api.BraintreeClient.sendPOST(BraintreeClient.java:218)
       at com.braintreepayments.api.ApiClient.tokenizeREST(ApiClient.java:64)
       at com.braintreepayments.api.PayPalInternalClient.tokenize(PayPalInternalClient.java:114)
       at com.braintreepayments.api.PayPalClient.onBrowserSwitchResult(PayPalClient.java:352)
       at com.braintreepayments.api.PayPalClient.deliverBrowserSwitchResultToListener(PayPalClient.java:284)
       at com.braintreepayments.api.PayPalClient.onBrowserSwitchResult(PayPalClient.java:279)
       at com.braintreepayments.api.PayPalLifecycleObserver$1.run(PayPalLifeCycleObserver.java:61)
       at android.os.Handler.handleCallback(Handler.java:809)
       at android.os.Handler.dispatchMessage(Handler.java:102)
       at android.os.Looper.loop(Looper.java:166)
       at android.app.ActivityThread.main(ActivityThread.java:7383)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:469)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:963)
```

**Comment by sshropshire** [comment]
Hi @tanmayyy thanks for using the Braintree SDK for Android. We've noticed a [number of issues](https://issuetracker.google.com/issues/158234058) with the security crypto library. We were notified that [active development has resumed](https://issuetracker.google.com/issues/215516227#comment6) on this library recently.

In the meantime, our best bet may be defensively guarding against a `SecurityException` being thrown within the SDK. Does this seem reasonable? There may be some undefined behavior when encrypted shared prefs fails, but we can prevent crashes this way.

**Comment by sshropshire** [comment]
@tanmayyy on second thought we may benefit from converting the unchecked exception internally and notifying the merchant SDK when a critical error occurs.

**Comment by jcnoriega** [comment]
@tanmayyy issue has been fixed in latest version 4.20.0. In order to avoid a crash, you need to catch BraintreeSharedPreferencesException. Let me know if it works and i'll close this issue!

**Comment by tanmayyy** [comment]
Thanks for the fix @sshropshire @jcnoriega ! We will upgrade the SDK and catch the excpetion indeed. Closing the issue, will re-open in case of a regression, have a good day!
